controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -277,7 +277,9 @@ func matchClusterConfiguration(kcp *controlplanev1.KubeadmControlPlane, machine
 	if ok {
 		machineClusterConfig := &kubeadmv1.ClusterConfiguration{}
 		// ClusterConfiguration annotation is not correct, only solution is to rollout.
-		if err := json.Unmarshal([]byte(machineClusterConfigStr), machineClusterConfig); err != nil {
+		// NOTE: it is required to handle a nil ClusterConfiguration (that is serialized into "null")
+		// see https://github.com/kubernetes-sigs/cluster-api/issues/3353
+		if err := json.Unmarshal([]byte(machineClusterConfigStr), &machineClusterConfig); err != nil {
```

Comment:
```suggestion
		// The call to json.Unmarshal has to take a pointer to the pointer struct defined above, otherwise we 
		// won't be able to handle a nil ClusterConfiguration (that is serialized into "null").
		// See https://github.com/kubernetes-sigs/cluster-api/issues/3353.
		if err := json.Unmarshal([]byte(machineClusterConfigStr), &machineClusterConfig); err != nil {
```